nifi-extension-bundles/nifi-github-bundle/nifi-github-extensions/src/main/java/org/apache/nifi/github/GitHubFlowRegistryClient.java

```diff
@@ -0,0 +1,633 @@
+        versionedFlowCoordinates.setBucketId(flowLocation.getBucketId());
+        versionedFlowCoordinates.setFlowId(flowLocation.getFlowId());
+        versionedFlowCoordinates.setVersion(createContentCommitSha);
+
+        flowSnapshot.getFlowContents().setVersionedFlowCoordinates(versionedFlowCoordinates);
+        flowSnapshot.getFlow().setBranch(branch);
+        flowSnapshot.getSnapshotMetadata().setBranch(branch);
+        flowSnapshot.getSnapshotMetadata().setVersion(createContentCommitSha);
+        flowSnapshot.setLatest(true);
+
+        // populate outgoing bucket references
+        updateBucketReferences(flowSnapshot, flowLocation.getBucketId());
+
+        // set back to the original id so that the returned snapshot is has the correct values from what was passed in
+        replaceGroupId(flowSnapshot.getFlowContents(), originalFlowContentsGroupId);
+        replacePosition(flowSnapshot.getFlowContents(), originalFlowContentsPosition);
+
+        return flowSnapshot;
+    }
+
+    @Override
+    public Set<RegisteredFlowSnapshotMetadata> getFlowVersions(final FlowRegistryClientConfigurationContext context, final FlowLocation flowLocation)
+            throws FlowRegistryException, IOException {
+        final GitHubRepositoryClient repositoryClient = getRepositoryClient(context);
+
+        final String branch = flowLocation.getBranch();
+        final String filePath = getSnapshotFilePath(flowLocation);
+
+        final Set<RegisteredFlowSnapshotMetadata> snapshotMetadataSet = new LinkedHashSet<>();
+        for (final GHCommit ghCommit : repositoryClient.getCommits(filePath, branch)) {
+            final RegisteredFlowSnapshotMetadata snapshotMetadata = createSnapshotMetadata(ghCommit, flowLocation);
+            if (REGISTER_FLOW_COMMENT.equals(snapshotMetadata.getComments())) {
+                continue;
+            }
+            snapshotMetadataSet.add(snapshotMetadata);
+        }
+        return snapshotMetadataSet;
+    }
+
+    @Override
+    public Optional<String> getLatestVersion(final FlowRegistryClientConfigurationContext context, final FlowLocation flowLocation) throws FlowRegistryException, IOException {
+        final GitHubRepositoryClient repositoryClient = getRepositoryClient(context);
+
+        final String branch = flowLocation.getBranch();
+        final String filePath = getSnapshotFilePath(flowLocation);
+
+        final List<GHCommit> commits = repositoryClient.getCommits(filePath, branch);
+        final String latestVersion = commits.isEmpty() ? null : commits.getFirst().getSHA1();
+        return Optional.ofNullable(latestVersion);
+    }
+
+    @Override
+    public String generateFlowId(final String flowName) {
+        return flowName.trim()
+                .replaceAll("\\s", "-") // replace whitespace with -
+                .replaceAll("[^a-zA-Z0-9-]", "") // replace all other invalid chars with empty string
+                .replaceAll("(-)\\1+", "$1"); // replace consecutive - with single -
```

Comment:
Can any type of collusion happen here? For example: I want a flow with the name of `flow 1` and `flow-1`, but the ids will collude. Do we want to prevent this kind of behaviour?

Reply:
The call to `registerFlow` will fail if a file with the same name already exists.